torch/_dynamo/output_graph.py

```diff
@@ -427,6 +427,17 @@ def init_ambient_guards(self):
             GlobalStateSource().make_guard(GuardBuilder.TORCH_FUNCTION_STATE)
         )
 
+        ci = torch._C._functorch.peek_interpreter_stack()
+        keys = (
+            torch._C._functorch.TransformType.Vmap,
+            torch._C._functorch.TransformType.Grad,
+            torch._C._functorch.TransformType.Jvp,
+        )
+        if ci is not None and ci.key() in keys:
+            self.guards.add(
+                GlobalStateSource().make_guard(GuardBuilder.FUNCTORCH_STACK_MATCH)
+            )
```

Comment:
NB: if there is anything on the stack, we should add a guard for it, not just these keys